src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+    {
+        _TaskbarProgressChangedHandlers(*this, nullptr);
+    }
+
+    bool ControlCore::HasSelection() const
+    {
+        return _terminal->IsSelectionActive();
+    }
+
+    bool ControlCore::CopyOnSelect() const
+    {
+        return _settings.CopyOnSelect();
+    }
+
+    std::vector<std::wstring> ControlCore::SelectedText(bool trimTrailingWhitespace) const
+    {
+        const auto bufferData = _terminal->RetrieveSelectedTextFromBuffer(trimTrailingWhitespace);
+        return bufferData.text;
+    }
+
+    ::Microsoft::Console::Types::IUiaData* ControlCore::GetUiaData() const
+    {
+        return _terminal.get();
+    }
+
+    // Method Description:
+    // - Search text in text buffer. This is triggered if the user click
+    //   search button or press enter.
+    // Arguments:
+    // - text: the text to search
+    // - goForward: boolean that represents if the current search direction is forward
+    // - caseSensitive: boolean that represents if the current search is case sensitive
+    // Return Value:
+    // - <none>
+    void ControlCore::Search(const winrt::hstring& text,
+                             const bool goForward,
+                             const bool caseSensitive)
+    {
+        if (text.size() == 0)
+        {
+            return;
+        }
+
+        const Search::Direction direction = goForward ?
+                                                Search::Direction::Forward :
+                                                Search::Direction::Backward;
+
+        const Search::Sensitivity sensitivity = caseSensitive ?
+                                                    Search::Sensitivity::CaseSensitive :
+                                                    Search::Sensitivity::CaseInsensitive;
+
+        ::Search search(*GetUiaData(), text.c_str(), direction, sensitivity);
+        auto lock = _terminal->LockForWriting();
+        if (search.FindNext())
+        {
+            _terminal->SetBlockSelection(false);
+            search.Select();
+            _renderer->TriggerSelection();
+        }
+    }
+
+    void ControlCore::SetBackgroundOpacity(const float opacity)
+    {
+        if (_renderEngine)
+        {
+            _renderEngine->SetDefaultTextBackgroundOpacity(::base::saturated_cast<float>(opacity));
+        }
+    }
+
+    // Method Description:
+    // - Asynchronously close our connection. The Connection will likely wait
+    //   until the attached process terminates before Close returns. If that's
+    //   the case, we don't want to block the UI thread waiting on that process
+    //   handle.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    winrt::fire_and_forget ControlCore::_asyncCloseConnection()
+    {
+        if (auto localConnection{ std::exchange(_connection, nullptr) })
+        {
+            // Close the connection on the background thread.
+            co_await winrt::resume_background();

```

Comment:
holy crap. since this gets called on destruction please write a huge comment in here to say **DO NOT INTERACT WITH THE CONTROLCORE AFTER THIS LINE** because it could have gone into the background after the destructor!